Jmx metrics insight blog (#2155)
open-telemetry/opentelemetry.io: content/en/blog/2023/jmx-metric-insight/index.md

---

> title: JMX Metric Insight

Reviewer: while the link title could stay short and sweet (JMX Metric Insight), I am wondering if here we should pick something more catchy, like "Gaining JMX Metric Insights with the OpenTelemetry Java Agent"

Author: Sounds good.

linkTitle: JMX Metric Insight
date: 2023-01-11
author: "[Samudraneel Dasgupta](https://github.com/Samudraneel24) (Cisco)"
---

[JMX](https://www.oracle.com/technical-resources/articles/javase/jmx.html) (Java
Management Extensions) is a technology that provides a way to manage and monitor
Java-based applications. Detailed information about the performance and resource
usage of the application can be derived from JMX metrics. This data can help you
identify trends or potential issues with the application and take action to
address them before they become serious problems. If there are any problems, we
can diagnose them with the help of metrics collected, and fine-tune the system
for optimal performance.

With the addition of the
[JMX Metric Insight](https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/jmx-metrics)

> module into the opentelemetry-java-instrumentation agent, we don't need to

Reviewer: ```suggestion
module into the [OpenTelemetry Javaagent](https://github.com/open-telemetry/opentelemetry-java-instrumentation), we don't need to
```

deploy a separate service just to collect JMX metrics for monitoring our
application. The agent can now natively collect and export metrics exposed by
application servers through local
[MBeans](https://docs.oracle.com/javase/tutorial/jmx/mbeans/index.html)
available within the instrumented application. The required MBeans and
corresponding metrics can be described using a YAML configuration file. The
individual metric configurations allow precise metric selection and
identification. JMX Metric Insight comes with a number of predefined
configurations containing curated sets of JMX metrics for popular application

> servers or frameworks, such as
> [ActiveMQ](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/jmx-metrics/javaagent/activemq.md),
> [Hadoop](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/jmx-metrics/javaagent/hadoop.md),
> [Jetty](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/jmx-metrics/javaagent/jetty.md),
> [Kafka Broker](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/jmx-metrics/javaagent/kafka-broker.md),
> [Tomcat](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/jmx-metrics/javaagent/tomcat.md)
> and
> [Wildfly](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/jmx-metrics/javaagent/wildfly.md).

Reviewer: ```suggestion
servers or frameworks, such as:

* [ActiveMQ](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/jmx-metrics/javaagent/activemq.md)
* [Hadoop](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/jmx-metrics/javaagent/hadoop.md)
* [Jetty](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/jmx-metrics/javaagent/jetty.md)
* [Kafka Broker](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/jmx-metrics/javaagent/kafka-broker.md)
* [Tomcat](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/jmx-metrics/javaagent/tomcat.md)
* [Wildfly](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/jmx-metrics/javaagent/wildfly.md)

```

You can also provide your own metric definitions, through one or more YAML

> files.

Reviewer: Add a pointer to the docs where people can learn how to write their own YAML file

    mapping:
      Count:
        metric: kafka.message.count
        type: counter
        desc: The number of messages received by the broker
        unit: "{messages}"
```

Each file can consist of multiple rules. Each rule can identify a set of one or
more MBeans, by the object name. In this example
`kafka.server:type=BrokerTopicMetrics,name=MessagesInPerSec` identifies a unique
MBean. We are interested in the attribute `Count` of this MBean, which is
specified under `mapping`. The metric reported will have the name
`kafka.message.count`, instrument type will be `counter` which indicates the
metric is a monotonic sum. It's unit will be `{messages}`. We have also provided
a description of the metric. This yaml segment is simple, to try out more
configuration options, you can head to the
[documentation](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/jmx-metrics/javaagent/README.md)
to understand and try out all the features of the module.

Lastly, if you feel some metrics are important to be in the predefined sets of
metrics, or in general if you have any idea for enhancement of the module, feel

> free to contribute to the repository.

Reviewer: add a link to the repository.

---
title: Gaining JMX Metric Insights with the OpenTelemetry Java Agent
linkTitle: JMX Metric Insight

> date: 2023-01-11

Reviewer: leaving this here: before going live/merging the PR we need to set the correct date.

> date: 2023-01-11

Reviewer 2: ```suggestion
date: 2023-01-17
```

```shell
zookeeper-server-start /usr/local/etc/kafka/zookeeper.properties
```

### Attach the OpenTelemetry java instrumentation agent

Before starting the kafka broker, attach the OpenTelemetry java instrumentation
agent to Kafka Broker by providing options in the KAFKA_OPTS environment
variable. You can download the latest release of the agent from
[here](https://github.com/open-telemetry/opentelemetry-java-instrumentation/releases).

```shell
export KAFKA_OPTS="-Dapplication.name=my-kafka-app
-Dotel.metrics.exporter=prometheus
-Dotel.exporter.prometheus.port=9464
-Dotel.service.name=my-kafka-broker
-Dotel.jmx.target.system=kafka-broker
-javaagent:/path/to/opentelemetry-javaagent.jar"
```

Now we can start the Kafka Broker:

```shell
kafka-server-start /usr/local/etc/kafka/server.properties
```

The Kafka broker should be up and running now. To test the installation, we can
create a topic and use the Kafka console producer and console consumer. Create
Kafka Topic:

```shell
kafka-topics --create --bootstrap-server localhost:9092 --replication-factor 1 --partitions 1 --topic my-test-topic
```

Start Kafka console producer to produce messages to the topic we created:

> ```shell
> kafka-console-producer --broker-list localhost:9092 --topic test
> >First message
> >Second message

Reviewer 3: If you're showing command output then use the `console` block language and include a command prompt, as shown below. Are the `>` actually part of the output?

```suggestion
```console
$ kafka-console-producer --broker-list localhost:9092 --topic test
>First message
>Second message
```

Author: Yes, the `>` are part of the Kafka ConsoleProducer output.

```

Now we will start the Kafka console consumer which will consume messages from
the topic from the beginning:

> ```shell
> kafka-console-consumer --bootstrap-server localhost:9092 --topic test --from-beginning
> First message
> Second message

Reviewer 3: ```suggestion
```console
$ kafka-console-consumer --bootstrap-server localhost:9092 --topic test --from-beginning
First message
Second message
```

backend of your choice. The full list of exporters and their configuration
options can be found
[here](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/autoconfigure/README.md#exporters).
For instance, you can export the metrics to an OTel collector using the OTLP
exporter, perform some processing and then consume the metrics on a backend of
your choice. In this example for the sake of simplicity, we are directly
exporting the metrics to Prometheus.

We will visualise the metrics on a Grafana dashboard using Prometheus as the
data source. In this demo, we’ll deploy Prometheus on Docker. We can create a
`prometheus.yml` file containing this minimal configuration:

```yaml
global:
  scrape_interval: 10s
  evaluation_interval: 10s

scrape_configs:
  - job_name: "my-kafka-broker"
    scrape_interval: 5s
    static_configs:
      - targets: ["host.docker.internal:9464"]
```

Then run the command below to deploy Prometheus on Docker:

```shell
docker run -d \
    -p 9090:9090 \
    -v path/to/prometheus.yml:/etc/prometheus/prometheus.yml \
    prom/prometheus
```

The Prometheus container should be running now. You can now navigate to

> [http://localhost:9090](http://localhost:9090) and explore the Prometheus

Reviewer 3: ```suggestion
<http://localhost:9090> and explore the Prometheus
```

dashboard. Here we are viewing the metric `kafka_request_count_total` on
Prometheus.

![A sample Kafka Broker metric shown on Prometheus](prometheus.png)

More installation options for Prometheus can be found
[here](https://prometheus.io/docs/prometheus/latest/installation/).

### View the metrics on a Grafana Dashboard

Now, we are going to visualise the Prometheus metrics in a Grafana dashboard. To
do that, first, pull the Grafana docker image using the following command:

```shell
docker run -d -p 3000:3000 grafana/grafana
```

> You can now navigate to [http://localhost:3000](http://localhost:3000) and

Reviewer 3: ```suggestion
You can now navigate to <http://localhost:3000> and
```

explore the Grafana home page. Click on Add Data Source and select Prometheus.

> Add the HTTP URL, default is `http://localhost:9090`. After that we can create

Reviewer 3: Can this be a link too (it might slightly easier to copy that way)?

```suggestion
Add the HTTP URL, default is <http://localhost:9090>. After that we can create
```
